File: core/FindBottomLeftTreeValue/FindBottomLeftTreeValue.java
package com.dbc.code;

public class FindBottomLeftTreeValue {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private int res = 0;
    private int resDeep = 0;
    private void dfs(TreeNode node, int deep) {
        if (node.left == null && node.right == null && deep > this.resDeep) {
            this.res = node.val;
            this.resDeep = deep;
        }
        if (node.left != null) dfs(node.left, deep + 1);
        if (node.right != null) dfs(node.right, deep + 1);
    }

    public int findBottomLeftValue(TreeNode root) {
        dfs(root, 1);
        return this.res;
    }
}

File: core/FindBottomLeftTreeValue/FindBottomLeftTreeValue.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def findBottomLeftValue(root: Optional[TreeNode]) -> int:
    res, res_deep = 0, 0
    def dfs(node, deep):
        nonlocal res, res_deep
        if not node.left and not node.right and deep > res_deep:
            res = node.val
            res_deep = deep
        if node.left:
            dfs(node.left, deep + 1)
        if node.right:
            dfs(node.right, deep + 1)

    dfs(root, 0)
    return res




